GeoffreyBooth-proposal.md

```diff
@@ -0,0 +1,78 @@
+Features that developers enjoy in ES2015 that are appropriate to implement in CoffeeScript, like `const`, should be implemented as time permits. These features should be implemented in a new compiler, assuming we build one; and possibly also in the old compiler too if such an implementation is easy and doesn’t break backward compatibility. Any feature implemented in both compilers should have identical syntax.
+
+### Modernizing Output
+
+Lastly, CoffeeScript’s output should be modernized as much as possible. Fat arrows should output as fat arrows, destructuring should output as destructuring, etc. But I can’t emphasize enough that this should be the **last priority**—it makes our source code cleaner, and debugging eventually easier when ES2015 features are supported natively in runtimes, but otherwise ES2015 output has little benefit for developers. That said, *new* features, especially ES2015+ ones, should definitely be output as ES2015+ code.
+
+## A New Compiler
+
+The current CoffeeScript 1.x compiler is extremely well commented, but it really needs those comments. Understanding it will bend your brain into knots. The original goal of [CoffeeScript Redux](https://github.com/michaelficarra/CoffeeScriptRedux) was to refactor CoffeeScript’s codebase, not to add new features.
+
+Don’t underestimate how hard it is to write your own language compiler. If it were easy, CoffeeScript Redux would have been finished; as it stands, that project was essentially abandoned a year ago.
+
+Also don’t underestimate how hard it is to add new features to the CoffeeScript 1.x compiler. If that were easy, someone else would’ve added modules and classes to it by now.
+
+The CoffeeScript community suffers from too many people with a lot of passion but little time to invest in development. For that reason, I consider a new-from-scratch compiler to be off the table. We’ll just end up with another Redux.
+
+So we have only two options for starting points for our new compiler:
+
+* The current CoffeeScript 1.x compiler.
+	* Pros: already supports all existing CoffeeScript code.
+	* Cons: complicated, difficult-to-understand codebase.
+* The CoffeeScript Redux compiler, probably from [this PR](https://github.com/michaelficarra/CoffeeScriptRedux/pull/344).
+	* Pros: better codebase, and the PR already supports many ES2015 features.
+	* Cons: we inherit Redux’s issues and unfinished work, including any incompatibilities with existing CoffeeScript code.
+
+I don’t have a recommendation as to one or the other. I invite contributors to both projects to chime in.
+
+## Workflow
+
+As to where we should put our new compiler and develop it, I have two ideas: 
+
+**Inside the current main [coffee-script repo](https://github.com/jashkenas/coffeescript).** Currently the repo is structured such that the code in `src` is compiled and output into `lib/coffee-script`. If we want to avoid merge conflicts with any pending PRs against the repo, we should leave those folders as they are.
```

Comment:
I think it would be better to migrate CoffeeScript Redux rather than fork it, so that we preserve its GitHub issues and pull requests. If we go the Redux route we’re basically taking on the task of finishing that project, so we might as well inherit its history.

I don’t think we should make any new repos under `coffeescript6`, other than perhaps for discussion. There is already an official `coffeescript` organization, which @jashkenas controls, so the new compiler effort should go there (or in the main repo if it’s easier to leave it under _jashkenas_ rather than move it under _coffeescript_).

The thing I really want to avoid is a splintering of resources, with teams pursuing competing efforts in separate repos. Our community suffers from a shortage of developers with time to donate, so whatever coding hours we can manage to organize need to be focused into one project.